<html>
<head>
  <title>snpminer</title>
</head>
<meta name="SNPFilter" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="./css/w3.css">
<link rel="stylesheet" href="./css/tooltip.css">
<style>
    * {
        box-sizing: border-box;
    }

    /* Create two equal columns that floats next to each other */
    .row{
        margin-bottom:8px;
    }

    .column {
        float: left;
        width: 50%;
        /*padding: 10px;*/
        /*line-height: 2.0;*/
    }

    /* Clear floats after the columns */
    .row:after {
        content: "";
        display: table;
        clear: both;
    }

    .btn{
        background: #5c7aff;
        padding:8px;
        width:10%;
        color:white;
        border-radius: 8px;
        font-weight:bold;
        text-transform: uppercase;
        align:"middle";
    }

    header{
        margin-bottom: 24px;
    }
    .textOverImage{
      width:300px;
      height:300px;
      float:left;
      margin:4px;
      transition:0.5s;
    }
    .textOverImage:after{
      content:attr(data-text)
    }

    .w3-container, .w3-panel {
      padding: 0.01em 95px;
    }
</style>
<body>


<header class="w3-container w3-blue">
    <h1>SNPtool</h1>
</header>

<div class="w3-container">
  <h2>
  SNPtool to filter and compare the SNPs
</h2>
<p style="font-size:16px; margin-left:20px;">
SNPtool is capable to filter out the SNPs (if this option is ticked) in your VCF files that do not meet the criteria specified in the filter options. This tool is also capable of comparing the SNPs in two or more VCFs (if this option is ticked) to obtain common SNPs in all VCFs or SNPs in a VCF that are not found in the other VCFs
</p>

<h3>How to use this webtool</h3>
<p>
  <ol>
    <li>Select and upload the reference sequence file (FASTA format) or file with reference sequence length data in two columns i.e. chromosome and length (.txt format)</li>
    <li>Select and upload one or more VCF format files</li>
    <li>Check the filter SNPs box, if you would like to filter the SNPs. The available options are listed upon checking the box. The values in the options are default options. Edit the values to your criteria. </li>
    <li>Check the compare SNPs box, if you would like to compare the SNPs in the VCF files. </li>
    <li>Click the Submit button</li>
    <li>A message will pop up once the processing completes. The result files are located in the folder "results"</li>
  </ol>
</p>

<h2>Let's use SNPtool now</h2>

    <hr>
    <form action="./cgi-bin/snpminer.py" method="get" enctype="multipart/form-data">
        <h3>Input reference sequence (FASTA) or length data (txt)</h3> <br/>

        <input type="file" name="reference_sequence" accept=".fa, .fasta, .txt"/> <br/></br>
        <hr>
        <h3>Input VCF format files</h3>

        Select and upload a one or multiple VCF files for filter and/or comparing SNPs using each add button </br>
        <div class="input_files_wrap">
            Upload one or more VCF files: </br></br>
            <input type="file" name="input_vcfs[]" accept=".vcf" id="upload" required multiple></br>
        </div>
      </br>
        <button type="button" class="add_field_button" style="border-radius:8px;background:lightskyblue;font-weight:bold">Add</button>
        <button type="button" class="remove_field_button" style="border-radius:8px;background:lightskyblue;font-weight:bold">Remove</button>

        <hr>
        <h3> Select what you want to do </h3>
        <input type="checkbox" id="filter-snps" name="filtersnps" onchange="filterOptionCheck()">Filter SNPs</input>

        <div class="tooltip">
          <img src="./images/help" alt="Filter SNPs." style="width:20px;height:20px;">
          <span class="tooltiptext">This option will filter out SNPs that don't meet the conditions set out in the options below</span>
        </div>

      </br>
        <input type="Checkbox" id="compare-snps" name="comparesnps" onchange="compareOptionCheck()">Compare snps</input>

        <div class="tooltip">
          <img src="./images/help" alt="Compare SNPs." style="width:20px;height:20px;">
          <span class="tooltiptext">This option will compare VCFs to find common SNPs or SNPs that are different at same position</span>
        </div>



        <div id="compare-toggle-options" style="display:none">
          <input type="checkbox" name="compare-common" id="compare-common" checked>Find Common SNPs</input>
          <input type="checkbox" name="compare-alternate" id="compare-alternate">Find Alternate SNPs</input>
        </br>Both 'Find Common SNPs' and 'Find Alternate SNPs' can be checked at the same time. Two separate outputs (tab-delimited) are generate for common SNPs and alternate SNPs.
        </div>

        <div id="filter-toggle-options" style="display:none">
          <hr>
          Filtering parameters: </br> </br>
          Note: Values for the parameters below are defaults. Please change them as you wish. </br></br>

          <hr>

          <h3>Genotyping</h3>

          Genotype: <input type="radio" id="Genotype" name="genotype" value="Homozygous" checked> Homozygous
          <input type="radio" id="Genotype" name="genotype" value="Heterozygous"> Heterozygous </br></br>

          <hr>

          <h3>Depth/Frequency:</h3>
          <div class="row">
              <div class="column">
                  Minimum variant allele frequency (FREQ):
              </div>
              <div class="column">
                  <input type="input" value=80 name="variant_allele_freq">
              </div>
          </div>


          <div class="row">
              <div class="column">
                  Minimum average per-sample depth of bases with Phred score >= 15 (ADP):
              </div>
              <div class="column">
                  <input type="input" value=3 name="per_sample_depth">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum number of samples called reference (wild-type) (WT):
              </div>
              <div class="column">
                  <input type="input" value=1 name="sample_called_reference">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Raw Read Depth as reported by SAMtools (SDP):
              </div>
              <div class="column">
                  <input type="input" value=3 name="raw_read_depth">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Quality Read Depth of bases with Phred score >= 15 (DP):
              </div>
              <div class="column">
                  <input type="input" value=10 name="quality_read_depth">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Depth of reference-supporting bases (reads1) (RD):
              </div>
              <div class="column">
                  <input type="input" value=3 name="ref_support_read_depth">
              </div>
          </div>


          <div class="row">
              <div class="column">
                  Minimum Depth of variant-supporting bases (reads2) (AD):
              </div>
              <div class="column">
                  <input type="input" value=3 name="variant_read_depth">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Depth of reference-supporting bases on forward strand (reads1plus) (RDF):
              </div>
              <div class="column">
                  <input type="input" value=3 name="depth_ref_support_forward_strand">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Depth of reference-supporting bases on reverse strand (reads1minus) (RDR):
              </div>
              <div class="column">
                  <input type="input" value=3 name="depth_ref_support_reverse_strand">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Depth of variant-supporting bases on forward strand (reads2plus) (ADF):
              </div>
              <div class="column">
                  <input type="input" value=3 name="depth_variant_support_forward_strand">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Depth of variant-supporting bases on reverse strand (reads2minus) (ADR):
              </div>
              <div class="column">
                  <input type="input" value=3 name="depth_variant_support_reverse_strand">
              </div>
          </div>


          <hr>

          <h3>Quality: </h3>
          <div class="row">
              <div class="column">
                  Minimum Genotype Quality (GQ):
              </div>
              <div class="column">
                  <input type="input" value=10 name="genotype_quality">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Average quality of reference-supporting bases (qual1) (RBQ):
              </div>
              <div class="column">
                  <input type="input" value=10 name="avg_qual_ref_support_bases">
              </div>
          </div>

          <div class="row">
              <div class="column">
                  Minimum Average quality of variant-supporting bases (qual2) (ABQ):
              </div>
              <div class="column">
                  <input type="input" value=10 name="avg_qual_variant_support_bases">
              </div>
          </div>

          <hr>

          <h3>P-value</h3>
          <div class="row">
              <div class="column">
                  P-value from Fisher's Exact Test (PVAL) less or equal to:
              </div>
              <div class="column">
                  <input type="number" value=0.05 id="pvalue" name="pvalue" min="0.01" max="1.0" step="0.01">
              </div>
          </div>

          <hr>
        </div>
        <hr>
        <input class="btn" type="submit" id="submit" value="submit" align="middle" onclick="validate_on_click_submit_button()"/>
    </form>

</div>

<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="./js/LineReader.js"></script>
<script src="./js/validate_vcf.js"></script>
<script src="./js/input_multiple_files.js"></script>

<script>
  function filterOptionCheck(){
    filterOptions = document.getElementById('filter-toggle-options')
    if (filterOptions.style.display === 'none'){
      filterOptions.style.display = 'inherit';
    } else {
      filterOptions.style.display = 'none'
    }
  }
  function compareOptionCheck(){
    compareOption = document.getElementById('compare-toggle-options')
    if (compareOption.style.display == 'none'){
      compareOption.style.display='inherit'
    } else {
      compareOption.style.display='none'
    }
  }
</script>
<script>
$(document).ready(function(){
    $('[data-toggle="tooltip"]').tooltip();
});
$(document).ready(function(){
    $('[data-toggle="popover"]').popover();
});
</script>


</body>
</html>
